Sources/Basic/TerminalController.swift

```diff
@@ -0,0 +1,226 @@
+*/
+
+import libc
+import func POSIX.getenv
+
+/// Tries to get the terminal width first using COLUMNS env variable and
+/// if that fails ioctl method testing on stdout stream.
+///
+/// - Returns: Current width of terminal if it was determinable.
+public func terminalWidth() -> Int? {
+    // Try to get from enviornment.
+    if let columns = POSIX.getenv("COLUMNS"), let width = Int(columns) {
+        return width
+    }
+
+    // Try determining using ioctl.
+    var ws = winsize()
+    if ioctl(1, UInt(TIOCGWINSZ), &ws) == 0 {
+        return Int(ws.ws_col)
+    }
+    return nil
+}
+
+/// Terminal color choices.
+public enum TermColor {
+    case noColor
+
+    case red
+    case green
+    case yellow
+    case cyan
+
+    case white
+    case black
+    case grey
+
+    /// Returns the color code which can be prefixed on a string to display it in that color.
+    private var string: String {
+        switch self {
+            case .noColor: return ""
+            case .red: return "\u{001B}[31m"
+            case .green: return "\u{001B}[32m"
+            case .yellow: return "\u{001B}[33m"
+            case .cyan: return "\u{001B}[36m"
+            case .white: return "\u{001B}[37m"
+            case .black: return "\u{001B}[30m"
+            case .grey: return "\u{001B}[30;1m"
+        }
+    }
+}
+
+/// A class to have better control on tty output streams: standard output and standard error.
+/// Allows operations like cursor movement and colored text output on tty.
+public final class TerminalController {
+
+    /// Pointer to output stream to operate on.
+    private var stream: LocalFileOutputByteStream
+
+    /// Width of the terminal.
+    public let width: Int
+
+    /// Code to clear the line on a tty.
+    private let clearLineString = "\u{001B}[2K"
+
+    /// Code to end any currently active wrapping.
+    private let resetString = "\u{001B}[0m"
+
+    /// Code to make string bold.
+    private let boldString = "\u{001B}[1m"
+
+    /// Constructs the instance if the stream is a tty.
+    public init?(stream: LocalFileOutputByteStream) {
+        // Make sure this file stream is tty.
+        guard !(isatty(fileno(stream.fp)) == 0) else {
+            return nil
+        }
+        width = terminalWidth() ?? 80 // Assume default if we are not able to determine.
+        self.stream = stream
+    }
+
+    /// Flushes the stream.
+    public func flush() {
+        try? stream.flush()
+    }
+
+    /// Clears the current line and moves the cursor to beginning of the line..
+    public func clearLine() {
+        stream <<< clearLineString + "\r"
+        flush()
+    }
+
+    /// Moves the cursor y columns up.
+    public func moveCursor(y: Int) {
+        stream <<< "\u{001B}[\(y)A"
+        flush()
+    }
+
+    /// Writes a string to the stream.
+    public func write(_ string: String, inColor color: TermColor = .noColor, bold: Bool = false) {
+        stream <<< wrap(string, inColor: color, bold: bold)
```

Comment:
Now that we have a stream, `wrap` can become something like `writeWrapped` or inlined here so that it avoids needing to create unnecessary temporary strings.